### ISSUE TITLE
Support different types for instance vs class variables

### ISSUE THREAD

**Issue by Ran4** [microsoft/pyright#3628]
**Is your feature request related to a problem? Please describe.**

Most (?) users of SQLAlchemy's ORM mode (https://docs.sqlalchemy.org/en/14/orm/quickstart.html) seems to have this problem:

You have defined an sql table using SQLAlchemy's ORM mode like this:

```python
from sqlalchemy import Column, Integer
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class Foo(Base):
    page: int = Column(Integer)

foo = Foo(page=34)
assert isinstance(foo.page, int)
assert isinstance(Foo.page, Column)  # note: different!
```

You use the fact that `Foo.page` is a Column when for example querying:

```python
page_one_and_forty_two = (
    db_session.query(Foo.page).filter(
        Foo.page.in_([1, 42])
    ).all()
)
```

This works, but pyright will complain about `Foo.page` being an int which doesn't have an `in_` method,
so the "correct" way of supressing pyrights warnings seems to be to do this very ugly hack:

```python
from typing import cast
from sqlalchemy import Column, Integer

# When defining the model:
class Foo(Base):
    page: int = cast(int, Column(Integer))

# When querying:
page_one_and_forty_two = (
    db_session.query(Foo.page).filter(
        cast(Column, Foo.page).in_([1, 42])
    ).all()
)
```

My current code base at work has thousands of these casts, and it's really ugly and cumbersome.

It gets extra confusing for readers when a field is optional, since your typing ends up as

```python
class Foo(Base):
    name: str | None = cast(str, Column(String, nullable=True))
```

as you can't cast into `str | None` but the instance variable is a `str | None`.

**Describe the solution you'd like**

Not sure, but some sort of function like cast that can tell Pyright that the instance variable type is different
than the class variable type (and/or vice versa).

Something like...

```python
class Foo(Base):
    page: int = instance_variable_cast(int, Column(Integer))
```

Even better would be if this could be done by the class itself, but I do not know enough about the internals of SQLAlchemy's declarative base to give a good suggestion.

**Comment by erictraut** [comment]
A type checker cannot know statically whether the reference to an attribute on an object will access an instance variable or a class variable. It depends on the history of that object — whether someone has written to the instance variable. That's just the way Python works. So I don't think it's feasible to support different types for instance or class variables.

You could start a conversation in the python/typing discussion forum to see if anyone else in the typing community has ideas on this front, but I think it's unlikely to gain traction. We certainly wouldn't create a pyright-specific mechanism to support this even if it were technical feasible to do so. Such a thing would need to be standardized.

Pyright (and other Python type checkers) already support the notion of a descriptor object stored as a class variable. These act like what you're describing, where the class variable is a descriptor object but references through an object go through the `__get__`, `__set__` or `__del__` methods on the descriptor, allowing it to return (or accept) different types.

I'm not that familiar with SQLAlchemy, but is `Column` a descriptor?
